[operator] Add upgrade hints and notes (#8433)
gardener/gardener: docs/concepts/operator.md

⚠️ Rotation of static `ServiceAccount` secrets is not supported since the `kube-controller-manager` does not enable the `serviceaccount-token` controller.

> ## Upgrading An Existing Gardener Landscape To `gardener-operator`

Reviewer: I see that we are not completely consistent with regards to casing in section titles. My preference would be to use lower case for filler words like "a/an" or "to". However, this is not done consistently in this document. Therefore, I leave it up to you.

> ## Upgrading An Existing Gardener Landscape To `gardener-operator`

Reviewer: Should we rather call it migrating instead of upgrading? The ownership/responsibility is migrated from something else, e.g. `garden-setup` to `gardener-operator`.

Since `gardener-operator` was only developed in 2023, six years after the Gardener project initiation, most users probably already have an existing Gardener landscape.
The most prominent installation procedure is [garden-setup](https://github.com/gardener/garden-setup), however experience shows that most community members have developed their own tooling for managing the garden cluster and the Gardener control plane components.

> Consequently, providing a general upgrade guide is not possible since the detailed steps vary heavily based on how the components were set up previously.

> > As a result, this section can only highlight the most important caveats and things to know, while the concrete upgrade steps must be figured out individually based on the existing installation.

Reviewer: Same as above: Maybe migration is the better word in this context.

Please make sure that you configure all your desired fields in the [`Garden` resource](#garden-resources).

### ETCD

> `gardener-operator` leverages `etcd-druid` for managing the `virtual-garden-etcd-main` and `virtual-garden-etcd-events`, similar to what happens for shoot clusters.

Reviewer: ```suggestion
`gardener-operator` leverages `etcd-druid` for managing the `virtual-garden-etcd-main` and `virtual-garden-etcd-events`, similar to how shoot cluster control planes are handled.
```

The `PersistentVolumeClaim` names differ slightly - for `virtual-garden-etcd-events` it's `virtual-garden-etcd-events-virtual-garden-etcd-events-0`, while for `virtual-garden-etcd-main` it's `main-virtual-garden-etcd-virtual-garden-etcd-main-0`.

The backup bucket must be created separately, and its name as well as the respective credentials must be provided via the `Garden` resource in `.spec.virtualCluster.etcd.main.backup`.

> 

Reviewer: So what is the ask here? Should an operator create a corresponding `PersistentVolumeClaim` and moving all existing data into the volume before starting the operator?
Same question about the backup.

In other words, do we want to provide more actionable documentation or leave the action items in the fineprint?

> ### `virtual-garden-kube-apiserver`

Reviewer: ```suggestion
### `virtual-garden-kube-apiserver` Deployment
```

    subjects:
    - kind: ServiceAccount
      name: virtual-garden-user
      namespace: kube-system
---
apiVersion: resources.gardener.cloud/v1alpha1
kind: ManagedResource
metadata:
  name: virtual-garden-access
  namespace: garden
spec:
  secretRefs:
  - name: managedresource-virtual-garden-access
```

The `shoot-access-virtual-garden` `Secret` will get a `.data.token` field which can be used to authenticate against the virtual garden cluster.
See also [this document](resource-manager.md#tokenrequestor-controller) for more information about the `TokenRequestor`.

### `gardener-apiserver`

Similar to the [`virtual-garden-kube-apiserver`](#virtual-garden-kube-apiserver), the `gardener-apiserver` also uses a few certificates and other credentials that should not change during the upgrade.
Again, you have to prepare the environment accordingly by leveraging the [secret's manager capabilities](../development/secrets_management.md#migrating-existing-secrets-to-secretsmanager).

- The existing ETCD Encryption Key `Secret` should be labeled with `secrets-manager-use-data-for-name=gardener-apiserver-etcd-encryption-key`.

> Please test your upgrade procedure thoroughly.

> > Note that in some cases it can be easier to set up a fresh landscape with `gardener-operator`, restore the ETCD data, switch the DNS records, and issue new credentials for all clients.

Reviewer: This seems good advise that should be generally followed, but it looks like it is just part of the `gardener-apiserver` section. Maybe it is a better idea to move it into the top section.
WDYT?
